File: backend/src/main/java/com/campussync/service/PostingQueryService.java
package com.campussync.service;

import com.campussync.dto.posting.PostingResponse;
import com.campussync.model.User;
import com.campussync.model.enums.PostingStatus;
import com.campussync.model.enums.SharingType;
import com.campussync.model.enums.TenantPreference;
import com.campussync.repository.PostingRepository;
import com.campussync.security.CurrentUserService;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * Listing discovery / browse (FR-3..FR-7, read side).
 * Reads run read-only so lazy associations map to DTOs before the session closes.
 */
@Service
@Transactional(readOnly = true)
public class PostingQueryService {

    private final PostingRepository postingRepository;
    private final CurrentUserService currentUserService;
    private final PostingSupport support;

    public PostingQueryService(PostingRepository postingRepository,
                               CurrentUserService currentUserService,
                               PostingSupport support) {
        this.postingRepository = postingRepository;
        this.currentUserService = currentUserService;
        this.support = support;
    }

    /**
     * Server-side discovery. The gender filter is applied here, NOT on the client:
     * a MALE seeker only ever receives MALE_ONLY + ANYONE listings, and vice-versa.
     * cityPrefix (city only), officeCampus (exact city+area) and tenantPreference are optional.
     */
    public List<PostingResponse> search(SharingType sharingType, String cityPrefix,
                                        String officeCampus, TenantPreference tenantPreference) {
        User me = currentUserService.get();
        return postingRepository.search(
                        PostingStatus.AVAILABLE, me.getGender(), sharingType,
                        support.blankToNull(cityPrefix), support.blankToNull(officeCampus), tenantPreference)
                .stream()
                .map(PostingResponse::from)
                .toList();
    }

    public PostingResponse getById(Long id) {
        return PostingResponse.from(support.findOrThrow(id));
    }

    public List<String> officeLocations() {
        return postingRepository.findDistinctOfficeCampuses();
    }
}

File: frontend/src/app/features/browse/posting-query.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Posting, SharingType, TenantPreference } from '../../models/models';

const API = 'http://localhost:8081/api/postings';

/** Listing discovery / browse — read side of the posting contract. */
@Injectable({ providedIn: 'root' })
export class PostingQueryService {
  private http = inject(HttpClient);

  /** Discovery. Gender filtering happens on the server; we just pass optional filters. */
  search(filters: { sharingType?: SharingType; city?: string; officeCampus?: string; tenantPreference?: TenantPreference } = {}): Observable<Posting[]> {
    let params = new HttpParams();
    if (filters.sharingType) params = params.set('sharingType', filters.sharingType);
    if (filters.city) params = params.set('city', filters.city);
    if (filters.officeCampus) params = params.set('officeCampus', filters.officeCampus);
    if (filters.tenantPreference) params = params.set('tenantPreference', filters.tenantPreference);
    return this.http.get<Posting[]>(API, { params });
  }

  locations(): Observable<string[]> {
    return this.http.get<string[]>(`${API}/locations`);
  }

  getById(id: number): Observable<Posting> {
    return this.http.get<Posting>(`${API}/${id}`);
  }
}
